examples/jsm/objects/BatchedMesh.js

```diff
@@ -0,0 +1,461 @@
+import {
+	BufferAttribute,
+	BufferGeometry,
+	DataTexture,
+	FloatType,
+	MathUtils,
+	Matrix4,
+	Mesh,
+	RGBAFormat
+} from 'three';
+
+const _identityMatrix = new Matrix4();
+const _zeroMatrix = new Matrix4().set(
+	0, 0, 0, 0,
+	0, 0, 0, 0,
+	0, 0, 0, 0,
+	0, 0, 0, 0
+);
+
+// Arbitary numbers
+const DEFAULT_MAX_GEOMETRY_COUNT = 256;
+const DEFAULT_MAX_VERTEX_COUNT = 1024 * 256;
+const DEFAULT_MAX_INDEX_COUNT = 1024 * 256;
+
+// Custom shaders
+const batchingParsVertex = `
+#ifdef BATCHING
+	attribute float id;
+	uniform highp sampler2D batchingTexture;
+	uniform int batchingTextureSize;
+	mat4 getBatchingMatrix( const in float i ) {
+		float j = i * 4.0;
+		float x = mod( j, float( batchingTextureSize ) );
+		float y = floor( j / float( batchingTextureSize ) );
+		float dx = 1.0 / float( batchingTextureSize );
+		float dy = 1.0 / float( batchingTextureSize );
+		y = dy * ( y + 0.5 );
+		vec4 v1 = texture2D( batchingTexture, vec2( dx * ( x + 0.5 ), y ) );
+		vec4 v2 = texture2D( batchingTexture, vec2( dx * ( x + 1.5 ), y ) );
+		vec4 v3 = texture2D( batchingTexture, vec2( dx * ( x + 2.5 ), y ) );
+		vec4 v4 = texture2D( batchingTexture, vec2( dx * ( x + 3.5 ), y ) );
+		mat4 bone = mat4( v1, v2, v3, v4 );
+		return bone;
+	}
+#endif
+`;
+
+const batchingbaseVertex = `
+#ifdef BATCHING
+	mat4 batchingMatrix = getBatchingMatrix( id );
+#endif
+`;
+
+const batchingnormalVertex = `
+#ifdef BATCHING
+	objectNormal = vec4( batchingMatrix * vec4( objectNormal, 0.0 ) ).xyz;
+	#ifdef USE_TANGENT
+		objectTangent = vec4( batchingMatrix * vec4( objectTangent, 0.0 ) ).xyz;
+	#endif
+#endif
+`;
+
+const batchingVertex = `
+#ifdef BATCHING
+	transformed = ( batchingMatrix * vec4( transformed, 1.0 ) ).xyz;
+#endif
+`;
+
+// @TODO: SkinnedMesh support?
+// @TODO: Move into the core. Can be optimized more with WEBGL_multi_draw.
```

Comment:
> Can be optimized more with WEBGL_multi_draw.

To me this is a big question ... doesn't WEBGL_multi_draw constrain the number of objects per batch quite a bit? Which could still be the better choice, but probably means a larger number of smaller batches, ideally co-located for culling, and more complex batch management...

I suspect it'll take us some time to figure that out, so I'm inclined to do our experimentation in addons and not rush this into core right away.

Reply:
Yeah, I don't think we should be in rush to move it to the core and use `WEBGL_multi_draw`, or even I'm not sure we really need to do them now. To clarify it, I added "future work" to the comment.